File: web/src/app/core/eeg/ring.ts
/**
 * A ring of samples held as one flat `Float32Array`.
 *
 * Flat and pre-allocated rather than an array of rows because the draw loop asks for the
 * newest few seconds several times a second, and allocating ~9,000 numbers each time
 * would produce a steady stream of garbage for the collector to chase.
 *
 * The wrap-around is the part worth testing: the arithmetic is the kind that is off by
 * one only once the buffer has actually wrapped, which is a minute into a session.
 */
export class SampleRing {
  private readonly data: Float32Array;
  private total = 0;

  constructor(
    private readonly capacity: number,
    readonly width: number,
  ) {
    if (capacity < 1) throw new Error('a sample ring needs a positive capacity');
    this.data = new Float32Array(capacity * width);
  }

  /** Rows currently available, which is `total` until the ring has wrapped. */
  get rows(): number {
    return Math.min(this.total, this.capacity);
  }

  /** Rows ever written, including those already overwritten. */
  get written(): number {
    return this.total;
  }

  push(rows: number[][]): void {
    for (const row of rows) {
      const base = (this.total % this.capacity) * this.width;
      for (let i = 0; i < this.width; i += 1) {
        this.data[base + i] = row[i] ?? 0;
      }
      this.total += 1;
    }
  }

  /**
   * Copy the newest `count` samples into `out`, oldest first, and return how many rows
   * were written. `out` must hold at least `capacity * width` values.
   */
  copyLatest(count: number, out: Float32Array): number {
    const n = Math.min(count, this.rows);
    if (n <= 0) return 0;
    const stride = this.width;
    const end = this.total % this.capacity || this.capacity;
    const start = end - n;
    if (start >= 0) {
      out.set(this.data.subarray(start * stride, end * stride), 0);
    } else {
      const head = -start;
      out.set(this.data.subarray((this.capacity - head) * stride), 0);
      out.set(this.data.subarray(0, end * stride), head * stride);
    }
    return n;
  }

  /** The newest few seconds, as rows of `width` values. Used by tests and previews. */
  latest(count: number): number[][] {
    const out = new Float32Array(this.capacity * this.width);
    const rows = this.copyLatest(count, out);
    const result: number[][] = [];
    for (let row = 0; row < rows; row += 1) {
      const slice: number[] = [];
      for (let column = 0; column < this.width; column += 1) {
        slice.push(out[row * this.width + column]);
      }
      result.push(slice);
    }
    return result;
  }
}

File: api/src/eeg_api/services/ring.py
"""A fixed-size ring buffer of samples.

The monitor holds only the display window in memory, so a long session cannot grow
the process. Pre-allocating the array (rather than a ``deque`` of rows) keeps the
per-tick cost a ``memcpy`` instead of thousands of small allocations.
"""

from __future__ import annotations

import numpy as np

from eeg_api.domain.models import FloatArray


class RingBuffer:
    """The newest ``capacity`` rows, oldest first, with O(1) append."""

    def __init__(self, capacity: int, width: int) -> None:
        if capacity < 1:
            raise ValueError("capacity must be positive")
        self.capacity = int(capacity)
        self.width = int(width)
        self._data = np.zeros((self.capacity, self.width), dtype=np.float64)
        self._total = 0

    def __len__(self) -> int:
        return min(self._total, self.capacity)

    @property
    def total(self) -> int:
        """Rows ever written, including those already overwritten."""
        return self._total

    def extend(self, rows: FloatArray) -> None:
        n = int(rows.shape[0])
        if n == 0:
            return
        if n >= self.capacity:
            self._data[:] = rows[-self.capacity :]
            self._total += n
            return
        start = self._total % self.capacity
        end = start + n
        if end <= self.capacity:
            self._data[start:end] = rows
        else:
            split = self.capacity - start
            self._data[start:] = rows[:split]
            self._data[: end - self.capacity] = rows[split:]
        self._total += n

    def latest(self, count: int) -> FloatArray:
        """The newest ``count`` rows, oldest first. Fewer if not yet available."""
        available = len(self)
        n = int(min(max(count, 0), available))
        if n == 0:
            return np.zeros((0, self.width), dtype=np.float64)
        end = self._total % self.capacity
        if end == 0:
            end = self.capacity
        start = end - n
        if start >= 0:
            return self._data[start:end].copy()
        return np.vstack([self._data[start:], self._data[:end]])

    def clear(self) -> None:
        self._data[:] = 0.0
        self._total = 0
